Add minimal set of macOS libc headers

kubkon opened the pull request:
Closes #5031

This PR adds a minimal set of macOS libc headers, ~~and updates README and adds LICENSE exception for macOS libc headers~~. Note that while technically with this PR it *is* possible to cross-compile any C binary to `x86_64-macos-gnu`, the resulting binary *will not* get loaded by `dyld` due to missing load command and improperly specified extern symbols (this is most likely a deficiency in `lld` which will hopefully soon-ish get fixed and/or we land our linker). In fact, I did a quick test and I manually modified the resultant binary (using a hex editor) to include the missing `LC_LOAD_DYLIB (libSystem.B.dylib)` load command, I've tweaked the ordinals in `Dynamic Loader Info` of Lazy and Nonlazy symbols to properly point at `libSystem.B.dylib`, and fixed the symbol table, and afterwards the binary was loaded properly by `dyld`. So we are not far off from a fully working setup.

In terms of a procedure for (re-)generating this set of headers, I've created a simplistic utility that can be run on a macOS to automatically obtain the relevant headers (many thanks to @andrewrk for suggesting this splendid hack!). Anyhow, you can find [the source of the utility here](https://github.com/kubkon/fetch-them-macos-headers).

Finally, as was rightly pointed out in #5031, some headers have missing/incomplete license info (like `stdint.h`), @andrewrk and I agreed this is an oversight on Apple's part and we will treat everything as APSL 2.0 licensed. This argument is ever stronger by the fact that [Apple released the libc as a whole under APSL 2.0 as evident in the official repo](https://opensource.apple.com/source/Libc/Libc-1353.100.2/).

I'm cc'ing everyone involved in #5031 to have a look and express their opinion about this PR: @aahmed-se, @daurnimator, @fengb, @judofyr, @andrewrk.

aahmed-se commented:
Looks good to me.

komuw commented on README.md:
```diff
@@ -122,5 +122,6 @@ However, such problems are best solved with social norms, not with software
 
-Therefore, Zig is available under the MIT (Expat) License, and comes with a
-humble request: use it to make software better serve the needs of end users.
+Therefore, Zig is available under the MIT (Expat) License with the exception
+of macOS libc headers which are licensed under Apple Public Source License Version 2.0,
+and comes with a humble request: use it to make software better serve the needs of end users.
```
I had to double read this part before it was clear what it meant. 

On my first reading, I incorrectly read it to mean; `macOS libc headers which are licensed under Apple Public Source License Version 2.0 and comes with a humble request ...`   
ie, I incorrectly read it as if the humble request is from Apple rather than been from Zig. 

For what it's worth, English is my 2nd language so this comment is more like a nitpick.

kubkon replied:
Not at all, this is some great feedback, thanks! Lemme redo this bit so that it's less confusing!

kubkon replied:
Rephrased! Lemme know if it makes more sense now :-)

komuw replied:
looks good. thanks.

Rocknest commented on README.md:
```diff
@@ -122,5 +122,7 @@ However, such problems are best solved with social norms, not with software
 
-Therefore, Zig is available under the MIT (Expat) License, and comes with a
-humble request: use it to make software better serve the needs of end users.
+Therefore, Zig is available under the MIT (Expat) License with the exception
+of macOS libc headers which are licensed under Apple Public Source License Version 2.0.
```
glibc is LGPL

kubkon replied:
That's a good point, thanks. Maybe while we're here we should add the missing exceptions to the master tree? @andrewrk what do you think?

kubkon replied:
FWIW I'm happy to make adjustments either in this or a subsequent PR if we want to keep the PR history clean.

andrewrk replied:
IMO just leave this alone for the PR, I'll do a follow up pass on the readme when merging, I want to rewrite that section entirely

kubkon replied:
SGTM! Lemme revert the README changes then.

zigazeljko commented:
> In terms of a procedure for (re-)generating this set of headers, I've created a simplistic utility that can be run on a macOS to automatically obtain the relevant headers.

Why exactly is this necessary? Can't we directly copy the headers from the libc sources released by Apple (https://opensource.apple.com/source/Libc/Libc-1353.100.2/)?

kubkon commented:
> > In terms of a procedure for (re-)generating this set of headers, I've created a simplistic utility that can be run on a macOS to automatically obtain the relevant headers.
> 
> Why exactly is this necessary? Can't we directly copy the headers from the libc sources released by Apple (https://opensource.apple.com/source/Libc/Libc-1353.100.2/)?

I thought so too initially, but it turns out Apple generates some headers automagically when building `libSystem.B.dylib` since they are platform specific... I tried to experiment with Apple's libc xcode project, but it wouldn't build at all. Too bad there's no docs, but that's no surprise!

daurnimator commented on lib/libc/include/x86_64-macos-gnu/signal.h:
```diff
@@ -0,0 +1,129 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. All advertising materials mentioning features or use of this software
```
@kubkon I don't think 4-clause BSD license is going to be good for us....

kubkon commented:
@andrewrk I've now reverted changes to README and I've removed LICENSE.macos-libc file for now (I figure you might have a better idea for handling the licensing exceptions than this).
